### ISSUE TITLE
[RFC] Ability to index multiple vector points per document

### ISSUE THREAD

**Issue by jobergum** [vespa-engine/vespa#24507]
With the growing interest in [approximate nearest neighbor search](https://docs.vespa.ai/en/approximate-nn-hnsw.html) with Vespa, there has also been increased demand for the ability to index multiple vectors per Vespa document. 

Motivation: 

- Transformer models have length limitations, so longer text needs to be chunked into paragraphs or sentence levels. The classic approach is splitting into separate Vespa documents, which causes redundancy of other fields, like the title, price, and availability. 
- Documents might have multi-valued vector properties. For example, an e-commerce product might be described with multiple images. [ColBERT](https://blog.vespa.ai/pretrained-transformer-language-models-for-search-part-3/) is a model where a text is represented using multi-vector representations. One vector per text token (subword)

## Schema 
Representing multiple vectors is best done using a [mixed tensor](https://docs.vespa.ai/en/tensor-user-guide.html) with one mapped dimension and a dense dimension. Using mixed mapped/dense allows for representing a variable number of vectors per document. 

For example, for an e-commerce search use case, we can encode all the product images using CLIP and index them for efficient text-to-image search. Similarly, the textual product data might be longer than transformer input length limitations (typically 512 subwords), so there is a need to have a chunked representation. 

```
schema product {
   document product {
      
      field title type string {..}
      field brand type string {..} 
      field available type bool {..}
      field categories type array<string> {..}
      field global_ctr_popularity type float {..}
      field text_chunks type array<string> {..}
      
      field image_embeddings type tensor<float>(image{}, x[768]) {
           indexing: attribute | index
      }
     
      field text_embeddings type tensor<float>(chunk{}, x[384]) {
          indexing: attribute | index
      }
   }

   rank-profile default {
      inputs {
        query(text_embedding) tensor<float>(x[384])
        query(image_embedding) tensor<float>(x[768])
      }
  }
}
```

A mixed tensor example 

```
tensor<float>(chunk{}, x[384]): {chunk1:[0.1,.....],chunk2:[0.02,...]}
```

## Query 
Given a user query `white cotton t-shirt for men with vertical grey stripes`, we can encode the query input using the text transformer and the text clip encoder and search in both embedding spaces using the same query request.  This example uses Vespa's ability to run inference; see [embeddings made easy](https://blog.vespa.ai/text-embedding-made-simple/). 

```json
{
 "yql": "select * from product where available = true and (({targetHits:10}nearestNeighbor(text_embeddings,text_embedding)) or ({targetHits:10}nearestNeighbor(image_embeddings,image_embedding)))",
"input.query(text_embedding)=embed(transformer, \"white cotton t-shirt for men with vertical grey stripes\")",
 "input.query(image_embedding)=embed(clip, \"white cotton t-shirt for men with vertical grey stripes\")"
}
```

This query searches both the text and image embedding space using OR. A simple `bool` filter [constrains](https://blog.vespa.ai/constrained-approximate-nearest-neighbor-search/) both nearest neighbor search operators. This costs less resource than doing two queries, as the filter execution is shared. 

 The query request is expected to expose at least 20 hits to the first-phase ranking. It's up to the ranking function to score the documents, for example, using a more advanced ranking model. 

## Ranking 
The `nearestNeighbor` query operator calculates a `rawScore` during the graph traversal, and the value depends on the `distance-metric` configured for the field.

This score can be accessed during ranking using the
either [closeness(dimension, name)](https://docs.vespa.ai/en/reference/rank-features.html#closeness(dimension,name)) or [distance(dimension, name)](https://docs.vespa.ai/en/reference/rank-features.html#distance(dimension,name)) rank features.

Today dimension can be `label` or `field`. Where `label` can be used if one has multiple nearestNeighbor query items searching the same field. (Batch query support). 

```
 rank-profile default {
      inputs {
        query(text_embedding) tensor<float>(x[384])
        query(image_embedding) tensor<float>(x[768])
      }
      first-phase {
         expression: closeness(field, text_embeddings) + closeness(field, image_embeddings)
      }
  }
```

A straightforward implementation would simply let closeness() return the score of the closest vector.
Since the graph traversal depends on a  proper`distance-metric`, we are limited to using the max score for the graph traversal/candidate retrieval. 

Still, users could use more advanced tensor aggregations in the first-phase ranking expression (e.g., using sum or average over the closeness, expressed as a tensor expression). But the retrieval phase is only limited to using max closeness/min distance. 

There is also a need to identify the closest vector, which vector made the document be retrieved. This could identify the "correct" image, when displaying the hit, or the "correct" text chunk. This could be done using a sub-feature, which is calculated in case the field is a mixed tensor (with multiple vectors):

```
closeness(field, text_embeddings).best
```
Which would return a tensor

```
{
    {chunk1:1}:1.0,
}
```
Indicating that the chunk with label 1 was the vector that caused the doc to be retrieved. This could then be used in either ranking or as part of `match-features` or `summary-features` to identify the best matching image/text chunk for display purposes. 

It could also be helpful to have all the similarity scores for all vectors: 

```
closeness(field, text_embeddings).scores 
```

Which would return a tensor with all the closeness/distance scores for all vectors:

```
tensor<float>(chunk{}): {chunk1: 0.3, chunk2:0.1, chunk3:0.2} 
```

In this case, users could use aggregates in the `first-phase` or `second-phase`. 
In this example, using the average of each closeness score: 

```
first-phase {
  expression: reduce(closeness(field, text_embeddings).scores , avg, chunk) 
} 
```

Using max aggregation: 
```
first-phase {
  expression: reduce(closeness(field, text_embeddings).scores , max, chunk) 
} 
```
Which would be equivalent to using 

```
first-phase {
  expression: closeness(field, text_embeddings)
} 
```

Similarly, there should be `distance(dimension, field).scores` / `distance(dimension, field).best`

# Feedback requested 
Any feedback on this feature is highly welcome.

**Comment by jobergum** [comment]
Another case not covered in the query example is the case where you have hybrid retrieval, combining the dense operators with sparse retrieval:

```
where available = true and (
  userQuery() or 
  ({targetHits:10}nearestNeighbor(text_embeddings,text_embedding)) or 
  ({targetHits:10}nearestNeighbor(image_embeddings,image_embedding))
)
```

where `userQuery()` is read from `query` and in Vespa 8 uses [weakAnd](https://docs.vespa.ai/en/using-wand-with-vespa.html). In this case, hits might be exposed to the `first-phase` ranking with no `closeness` or `rawScore` scores. Should we, in this case, calculate it on-demand so that the end user doesn't need to write another tensor expression to cover this case?

**Comment by jobergum** [comment]
Alternative names


- `closeness(field, text_embeddings).closest` versus `closeness(field, text_embeddings).best`
- `closeness(field, text_embeddings).all` versus `closeness(field, text_embeddings).scores` 

- `distance(field, text_embeddings).closest` versus `distance(field, text_embeddings).best`
- `distance(field, text_embeddings).all` versus `distance(field, text_embeddings).scores` 

The first feature returns a single tensor with the label of the closest set to 1.0. 

The second feature returns a tensor with all labels and the distances/closeness float score.

**Comment by peterdm** [comment]
Support for multiple vectors per document is needed to avoid having to create artificially granular documents!  (E.g. Instead of BlogPostSentence, it would be easier to index BlogPost (but with a list of sentence embeddings contained inside))

**Comment by bratseth** [comment]
Yes, although for text documents in general the best option might be to map it to N vectors representing the concepts discussed (a "vector cloud"), rather than one vector per sentence.

**Comment by roope-kvp** [comment]
This sounds very much like what I need to index longer texts.  I need to rank higher texts that generally match the query, but also point out which paragraphs in particular match best.

**Comment by jobergum** [comment]
Multi-vector indexing is available from Vespa 8.144.19.
https://blog.vespa.ai/semantic-search-with-multi-vector-indexing/
